File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.java
class Solution {
    public int helper(int i, int[] prices, boolean isBuying, Map<String, Integer> visited) {
        if (i >= prices.length)
        return 0;

        String curState = String.valueOf(i) + String.valueOf(isBuying);
        if (visited.containsKey(curState)) return visited.get(curState);

        if (isBuying) {
            int buy = helper(i + 1, prices, !isBuying, visited) - prices[i];
            int cooldown = helper(i + 1, prices, isBuying, visited);
            visited.put(curState, Math.max(buy, cooldown));
        } else {
            int sell = helper(i + 2, prices, !isBuying, visited) + prices[i];
            int cooldown = helper(i + 1, prices, isBuying, visited);
            visited.put(curState, Math.max(sell, cooldown));
        }

        return visited.get(curState);
    }

    public int maxProfit(int[] prices) {
        return helper(0, prices, true, new HashMap());
    }
}

File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        def dfs(i, choice):
            if (i == len(prices)):
                return 0

            '''
            Buy(0) -> profit -= prices[i]

            Sell(1) -> profit += prices[i]
                 -> profit = profit
            '''

            holdProfit = dfs(i + 1, choice)
            if (choice == 0): 
                # buy
                buyProfit = dfs(i + 1, 1) - prices[i]
                return max(buyProfit, holdProfit)
            else:
                # sell
                sellProfit = dfs(i + 1, 0) + prices[i]
                return max(holdProfit, sellProfit)
                
        return dfs(0, 0)